src/NuGet.Services.AzureSearch/Auxiliary2AzureSearch/DataSetComparer.cs

```diff
@@ -0,0 +1,145 @@
+﻿// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System;
+using System.Collections.Generic;
+using System.Diagnostics;
+using System.Linq;
+using Microsoft.Extensions.Logging;
+
+namespace NuGet.Services.AzureSearch.Auxiliary2AzureSearch
+{
+    public class DataSetComparer : IDataSetComparer
+    {
+        private static readonly string[] EmptyStringArray = new string[0];
+
+        private readonly IAzureSearchTelemetryService _telemetryService;
+        private readonly ILogger<DataSetComparer> _logger;
+
+        public DataSetComparer(
+            IAzureSearchTelemetryService telemetryService,
+            ILogger<DataSetComparer> logger)
+        {
+            _telemetryService = telemetryService ?? throw new ArgumentNullException(nameof(telemetryService));
+            _logger = logger ?? throw new ArgumentNullException(nameof(logger));
+        }
+
+        public SortedDictionary<string, string[]> CompareOwners(
+            SortedDictionary<string, SortedSet<string>> oldData,
+            SortedDictionary<string, SortedSet<string>> newData)
+        {
+            // Use ordinal comparison to allow username case changes to flow through.
+            var stopwatch = Stopwatch.StartNew();
+            var result = CompareData(
+                oldData,
+                newData,
+                "package ID",
+                "owners",
+                StringComparer.Ordinal);
+
+            stopwatch.Stop();
+            _telemetryService.TrackOwnerSetComparison(oldData.Count, newData.Count, result.Count, stopwatch.Elapsed);
+
+            return result;
+        }
+
+        public SortedDictionary<string, string[]> ComparePopularityTransfers(
+            SortedDictionary<string, SortedSet<string>> oldData,
+            SortedDictionary<string, SortedSet<string>> newData)
+        {
+            // Ignore case changes in popularity transfers.
+            var stopwatch = Stopwatch.StartNew();
+            var result = CompareData(
+                oldData,
+                newData,
+                "package ID",
+                "popularity transfers",
+                StringComparer.OrdinalIgnoreCase);
+
+            stopwatch.Stop();
+            _telemetryService.TrackPopularityTransfersSetComparison(oldData.Count, newData.Count, result.Count, stopwatch.Elapsed);
+
+            return result;
+        }
+
+        private SortedDictionary<string, string[]> CompareData(
+            SortedDictionary<string, SortedSet<string>> oldData,
+            SortedDictionary<string, SortedSet<string>> newData,
+            string keyName,
+            string valuesName,
+            StringComparer valuesComparer)
+        {
```

Comment:
This logic is almost the same as the previous [`OwnerSetComparer.Compare` method](https://github.com/NuGet/NuGet.Services.Metadata/blob/91b715a871b538389d3f1f14717fed9eec2dd2e4/src/NuGet.Services.AzureSearch/Auxiliary2AzureSearch/OwnerSetComparer.cs#L30).